lib/driver.js

```diff
@@ -184,8 +194,29 @@ class BaseDriver extends MobileJsonWireProtocol {
 
     if (!_.contains(validStrats, strategy)) {
       throw new errors.InvalidSelectorError(`Locator Strategy "${strategy}" is not supported for this session`);
+    }
+  }
+
+  /*
+   * Restart the session with the original caps,
+   * preserving the timeout config.
+   */
+  async reset () {
+    log.debug("Resetting app mid-session");
+    log.debug("Running generic full reset");
+    let currentConfig = {};
+    for (let k of ['implicitWaitMs', 'newCommandTimeoutMs', 'sessionId']) {
+      currentConfig[k] = this[k];
+    }
+    await this.deleteSession();
+    log.debug("Restarting app");
+    await this.createSession(this.caps);
+
+    for (let [k,v] of _.pairs(currentConfig)) {
+      this[k] = v;
```

Comment:
It could work. We do need it in a few places as you said but maybe there's a more elegant way. Would be kind of nice to get that state out of the driver. That means we need to do session state checking in the webserver somehow (or mjsonwp)